### ISSUE TITLE
Varying results for equavalent code execution

### ISSUE THREAD

**Issue by eveselov** [Microsoft/CNTK#1403]
I must be missing something obvious, but why the following code gives different results for repeated executions. I can only guess that there is some hidden randomness in the my model construction code...
To reproduce copy paste the following code into jupyter notebook and hit shift-enter several times. You will see different results produced. This is not what I expect here because the model seems fully deterministic to me. This happens only if the model itself  is reconstructed; if you only rerun the model evaluation, it's deterministic as expected. I am concerned that I might be missing something important about the api behavior...

from cntk import Axis
from cntk.learner import momentum_sgd, momentum_as_time_constant_schedule, learning_rate_schedule, UnitType
from cntk.ops import input_variable, placeholder_variable, combine, sequence, past_value
from cntk.blocks import LSTM
input_dim = 2
hid_dim = 6
output_dim = 1
input = input_variable(shape=(input_dim), dynamic_axes=[Axis.default_batch_axis(), Axis('inputAxis')])
dh = placeholder_variable(shape=(hid_dim), dynamic_axes=input.dynamic_axes)
dc = placeholder_variable(shape=(hid_dim), dynamic_axes=input.dynamic_axes)
lstm_cell = LSTM(hid_dim)(input, (dh, dc))
lstm_cell.replace_placeholders({ dh: past_value(lstm_cell.outputs[0]).output, dc: past_value(lstm_cell.outputs[1]).output })
model = combine([lstm_cell.outputs[0]])
X = [[0, 1.0], [0, 2.0], [0, 3.0], [0, 4.0], [0, 5.0], [0, 6.0], [0, 7.0]]
model.eval({input: X})

**Comment by eveselov** [comment]
Well I figured this out. It have taken less time :).
Apparently LSTM(hid_dim)(input, (dh, dc) creates a structure with three parameters embedded in it.
Two of these parameters are randomly populated in the moment of this function creation. This is a source of mysterious randomness. When I added a couple of explicit parameters assignments like:
     model.parameter[1].value = np.array([[...]]) 
the behavior became deterministic as I expected.

The overall feedback from my as a new learner is that it was a bit unreasonable: too smart. The random parameter initialization is needed to help training. So strictly speaking it is unexpected for pure function definition. It would save my time if this randomness was at least more explicit in function calls or in tutorials or documentation (did I miss something?).

Anyway, thanks a lot for a great package.
